raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -883,6 +923,13 @@ private boolean handleVoteResponse(
         VoteResponseData response = (VoteResponseData) responseMetadata.data();
         Errors topLevelError = Errors.forCode(response.errorCode());
         if (topLevelError != Errors.NONE) {
+            if (topLevelError == Errors.UNSUPPORTED_VERSION && quorum.isProspective()) {
+                logger.warn("Prospective received unsupported version error in vote response in epoch {}, " +
+                        "transitioning to Candidate state immediately since entire quorum does not support PreVote.",
+                    quorum.epoch());
```

Comment:
This is not a warning. This is a valid state or condition. We should be able to log this message at INFO level. It should be rare because there are backoff/timeout logic in Candidate, Follower and Unattached which limit how quickly a replica transitions to Prospective. What do you think?

Reply:
> This is a valid state or condition.
I see your point here, there's nothing 'incorrect' about this happening. 
> It should be rare because there are backoff/timeout logic in Candidate, Follower and Unattached which limit how quickly a replica transitions to Prospective.
But I'm wondering about the case where a controller quorum is left partially upgraded on accident. Would having a warning log make this situation more discoverable?

Reply:
> But I'm wondering about the case where a controller quorum is left partially upgraded on accident. Would having a warning log make this situation more discoverable?

Yeah. This would be interesting to monitor but cluster are not really monitored by looking at the log. Clusters are monitored by collecting and comparing metrics.

This is an issue beyond KRaft and should be solved holistically. Kafka could have a metrics that fires if the API versions and supported features don't match across all of the replicas. This metric would require a KIP to implement.

Having said that, we should keep this message but it should not be logged at WARN. It should be logged at INFO.